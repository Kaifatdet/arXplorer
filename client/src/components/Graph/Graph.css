.data-container {
  margin: 0 auto;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 80vh;
  background-color: rgb(178, 178, 178, 0.1);
  box-shadow: 0 0 5px 1px #3333;
  margin-bottom: 2rem;
}

.graph-svg {
  width: 100%;
  height: 100%;
}

.graph-clean {
  display: flex;
  flex-direction: row-reverse;
  align-items: center;
  position: absolute;
  top: 80vh;
  right: 2rem;
}

.graph-icon-clean {
  font-size: 3rem;
  color: var(--main-bg-dark);
}

.graph-icon-clean:hover {
  color: var(--red);
}

.kill-text {
  display: none;
  margin-right: 5px;
}

.graph-icon-clean:hover + .kill-text {
  display: block;
}

.graph-empty-placeholder {
  position: absolute;
  top: 10rem;
  left: 50%;
  width: 20vw;
  text-align: center;
}

#search-redirect-link {
  text-decoration: none;
  color: rgb(0, 179, 255);
}

#search-redirect-link:hover {
  text-decoration: underline;
}