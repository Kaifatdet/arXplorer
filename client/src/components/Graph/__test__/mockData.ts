export const mockData = {
  nodes: [
    {
      id: 'Darwin Chang',
      weight: 16,
      group: 1,
      cat: 'hep-ph',
      cat_name: 'High Energy Physics - Phenomenology',
    },
    {
      id: 'Marc Schumann',
      weight: 0,
      group: 2,
      cat: 'astro-ph.IM',
      cat_name: 'Instrumentation and Methods for Astrophysics',
    },
    {
      id: 'Laura Baudis',
      weight: 0,
      group: 2,
      cat: 'astro-ph.IM',
      cat_name: 'Instrumentation and Methods for Astrophysics',
    },
    {
      id: 'W. -Y. Keung',
      weight: 1,
      group: 1,
      cat: 'hep-ph',
      cat_name: 'High Energy Physics - Phenomenology',
    },
    {
      id: 'Ernest Ma',
      weight: 1,
      group: 1,
      cat: 'hep-ph',
      cat_name: 'High Energy Physics - Phenomenology',
    },
    {
      id: 'A. Zee',
      weight: 1,
      group: 1,
      cat: 'hep-ph',
      cat_name: 'High Energy Physics - Phenomenology',
    },
    {
      id: 'R. N. Mohapatra',
      weight: 1,
      group: 1,
      cat: 'hep-ph',
      cat_name: 'High Energy Physics - Phenomenology',
    },
    {
      id: 'Darwin Bautista',
      weight: 1,
      group: 3,
      cat: 'cs.LG',
      cat_name: 'Learning',
    },
    {
      id: 'Raimarc Dionido',
      weight: 1,
      group: 3,
      cat: 'cs.LG',
      cat_name: 'Learning',
    },
    {
      id: 'Oscar Darwin',
      weight: 1,
      group: 4,
      cat: 'cs.LO',
      cat_name: 'Logic in Computer Science',
    },
    {
      id: 'Stefan Kiefer',
      weight: 1,
      group: 4,
      cat: 'cs.LO',
      cat_name: 'Logic in Computer Science',
    },
    {
      id: 'Mladen Georgiev',
      weight: 1,
      group: 5,
      cat: 'physics.ao-ph',
      cat_name: 'Atmospheric and Oceanic Physics',
    },
    {
      id: 'Jai Singh',
      weight: 1,
      group: 5,
      cat: 'physics.ao-ph',
      cat_name: 'Atmospheric and Oceanic Physics',
    },
    {
      id: 'Dung-Hai Lee',
      weight: 1,
      group: 6,
      cat: 'cond-mat.str-el',
      cat_name: 'Strongly Correlated Electrons',
    },
    {
      id: 'Wai-Yee Keung',
      weight: 4,
      group: 1,
      cat: 'hep-ph',
      cat_name: 'High Energy Physics - Phenomenology',
    },
    {
      id: 'Ivan Phillips',
      weight: 2,
      group: 1,
      cat: 'hep-ph',
      cat_name: 'High Energy Physics - Phenomenology',
    },
    {
      id: 'Jiang Liu',
      weight: 2,
      group: 1,
      cat: 'hep-ph',
      cat_name: 'High Energy Physics - Phenomenology',
    },
    {
      id: 'Palash B. Pal',
      weight: 2,
      group: 1,
      cat: 'hep-ph',
      cat_name: 'High Energy Physics - Phenomenology',
    },
    {
      id: 'Palash Pal',
      weight: 2,
      group: 1,
      cat: 'hep-ph',
      cat_name: 'High Energy Physics - Phenomenology',
    },
    {
      id: 'Xiao-Gang He',
      weight: 2,
      group: 1,
      cat: 'hep-ph',
      cat_name: 'High Energy Physics - Phenomenology',
    },
    {
      id: 'Sandip Pakvasa',
      weight: 2,
      group: 1,
      cat: 'hep-ph',
      cat_name: 'High Energy Physics - Phenomenology',
    },
    {
      id: 'Shih-Chang Lee',
      weight: 2,
      group: 1,
      cat: 'hep-ph',
      cat_name: 'High Energy Physics - Phenomenology',
    },
    {
      id: 'Paul Turcotte',
      weight: 2,
      group: 1,
      cat: 'hep-ph',
      cat_name: 'High Energy Physics - Phenomenology',
    },
    {
      id: 'Chuan-Hung Chen',
      weight: 1,
      group: 1,
      cat: 'hep-ph',
      cat_name: 'High Energy Physics - Phenomenology',
    },
    {
      id: 'David Bowser-Chao',
      weight: 2,
      group: 1,
      cat: 'hep-ph',
      cat_name: 'High Energy Physics - Phenomenology',
    },
  ],
  links: [
    { source: 'Darwin Chang', target: 'W. -Y. Keung' },
    { source: 'Darwin Chang', target: 'Ernest Ma' },
    { source: 'Darwin Chang', target: 'A. Zee' },
    { source: 'Darwin Chang', target: 'R. N. Mohapatra' },
    { source: 'Darwin Chang', target: 'Dung-Hai Lee' },
    { source: 'Darwin Chang', target: 'Wai-Yee Keung' },
    { source: 'Darwin Chang', target: 'Ivan Phillips' },
    { source: 'Darwin Chang', target: 'Jiang Liu' },
    { source: 'Darwin Chang', target: 'Palash B. Pal' },
    { source: 'Darwin Chang', target: 'Palash Pal' },
    { source: 'Darwin Chang', target: 'Xiao-Gang He' },
    { source: 'Darwin Chang', target: 'Sandip Pakvasa' },
    { source: 'Darwin Chang', target: 'Shih-Chang Lee' },
    { source: 'Darwin Chang', target: 'Paul Turcotte' },
    { source: 'Darwin Chang', target: 'Chuan-Hung Chen' },
    { source: 'Darwin Chang', target: 'David Bowser-Chao' },
    { source: 'Darwin Bautista', target: 'Raimarc Dionido' },
    { source: 'Oscar Darwin', target: 'Stefan Kiefer' },
    { source: 'Mladen Georgiev', target: 'Jai Singh' },
    { source: 'Wai-Yee Keung', target: 'Ivan Phillips' },
    { source: 'Wai-Yee Keung', target: 'Palash Pal' },
    { source: 'Wai-Yee Keung', target: 'David Bowser-Chao' },
    { source: 'Jiang Liu', target: 'Palash B. Pal' },
    { source: 'Xiao-Gang He', target: 'Sandip Pakvasa' },
    { source: 'Shih-Chang Lee', target: 'Paul Turcotte' },
  ],
};
