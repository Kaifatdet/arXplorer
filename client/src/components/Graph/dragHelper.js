'use strict';

import { drag, scaleOrdinal, schemePaired } from 'd3';

export const dragFunc = (simulation) => {
  function dragstarted(event) {
    if (!event.active) simulation.alphaTarget(0.5).restart();
    event.subject.fx = event.subject.x;
    event.subject.fy = event.subject.y;
  }

  function dragged(event) {
    event.subject.fx = event.x;
    event.subject.fy = event.y;
  }

  function dragended(event) {
    if (!event.active) simulation.alphaTarget(0);
    // event.subject.fx = null;
    // event.subject.fy = null;
    event.subject.fx = event.x;
    event.subject.fy = event.y;
  }

  return drag()
    .on('start', dragstarted)
    .on('drag', dragged)
    .on('end', dragended);
};

export const backgroundDrag = (selection, width, height, offsetX, offsetY) => {
  function dragstarted() {
    return undefined;
  }

  function dragged(e) {
    offsetX -= e.dx;
    offsetY -= e.dy;
    selection.attr('viewBox', [
      0 + offsetX,
      0 + offsetY,
      width + offsetX,
      height + offsetY,
    ]);
  }

  function dragended() {
    return undefined;
  }

  return drag()
    .on('start', dragstarted)
    .on('drag', dragged)
    .on('end', dragended);
};

export const color = () => {
  const scale = scaleOrdinal(schemePaired);
  return (d) => scale(d.group);
};
