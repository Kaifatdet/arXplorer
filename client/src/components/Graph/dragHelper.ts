import {
  D3DragEvent,
  drag,
  scaleOrdinal,
  schemePaired,
  Simulation,
  Selection,
} from 'd3';
import { GraphNode } from '../../types';

export const dragFunc = (simulation: Simulation<GraphNode, undefined>) => {
  function dragStarted(
    event: D3DragEvent<SVGSVGElement, GraphNode, GraphNode>
  ) {
    if (!event.active) simulation.alphaTarget(0.5).restart();
    event.subject.fx = event.subject.x;
    event.subject.fy = event.subject.y;
  }

  function dragged(event: D3DragEvent<SVGSVGElement, GraphNode, GraphNode>) {
    event.subject.fx = event.x;
    event.subject.fy = event.y;
  }

  function dragEnded(event: D3DragEvent<SVGSVGElement, GraphNode, GraphNode>) {
    if (!event.active) simulation.alphaTarget(0);
    event.subject.fx = event.x;
    event.subject.fy = event.y;
  }

  return drag()
    .on('start', dragStarted)
    .on('drag', dragged)
    .on('end', dragEnded);
};

export const backgroundDrag = (
  selection: Selection<SVGSVGElement, GraphNode, HTMLElement, undefined>,
  offsetX: number,
  offsetY: number
) => {
  function dragStarted() {
    return undefined;
  }

  function dragged(e: D3DragEvent<SVGSVGElement, GraphNode, GraphNode>) {
    offsetX -= e.dx;
    offsetY -= e.dy;
    selection
      .selectAll('g.nodes')
      .attr('transform', `translate(${-offsetX}, ${-offsetY})`);
  }

  function dragEnded() {
    return undefined;
  }

  return drag()
    .on('start', dragStarted)
    .on('drag', dragged)
    .on('end', dragEnded);
};

export const color = () => {
  const scale = scaleOrdinal(schemePaired);
  return (d: any) => scale(d.group);
};
