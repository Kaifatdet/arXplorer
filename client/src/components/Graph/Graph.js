'use strict';

import React, { useRef, useEffect, useState } from 'react';
import './Graph.css';
import drawGraph from './drawGraph';
import { select } from 'd3';
import RightSidebar from '../RightSidebar';

function Graph({ graphData, handleGraphExpand, authorDict }) {
  const [selected, setSelected] = useState('');
  const svgRef = useRef();
  const dimensions = {
    height: 600,
    width: 1280,
  };

  const handleClick = (input) => {
    setSelected(input);
  };

  const handleExpandClick = () => {
    handleGraphExpand(selected);
  };

  const toggleSelected = () => {
    setSelected('');
  };

  const extractCategories = (data) => {
    let cats = [];
    data.forEach((n) => {
      if (cats.filter((el) => el.name === n.cat_name).length === 0) {
        cats.push({
          name: n.cat_name,
          group: n.group,
        });
      }
    });
    return cats;
  };

  useEffect(() => {
    const svg = select(svgRef.current);
    drawGraph(svg, graphData, dimensions, handleClick, extractCategories);
  }, [graphData]);

  return (
    <div className="graph-container">
      <div className="data-container">
        <svg ref={svgRef}></svg>
      </div>
      <RightSidebar
        selected={selected}
        handleExpandClick={handleExpandClick}
        authorDict={authorDict}
        toggleSelected={toggleSelected}
      />
    </div>
  );
}

export default Graph;
