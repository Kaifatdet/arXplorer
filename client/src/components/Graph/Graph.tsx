import React, { useRef, useEffect, useState, FunctionComponent } from 'react';
import { select } from 'd3';
import { drawGraph } from './drawGraph';
import RightSidebar from '../RightSidebar';
import TinySearchBar from '../TinySearchBar';
import GraphErrorHandler from '../GraphErrorHandler';
import { Dictionary, Dimensions, GraphData, GraphNode } from '../../types';
import './Graph.css';

interface GraphProps {
  dimensions: Dimensions;
  graphData: GraphData | null;
  handleGraphExpand: (author: string) => Promise<void>;
  authorDict: Dictionary;
  selectedAuthor: string;
  setSelectedAuthor: React.Dispatch<React.SetStateAction<string>>;
  setSelectedArticle: React.Dispatch<React.SetStateAction<string>>;
  removeSelectedAuthor: (author: string) => void;
  handleQuickSearch: (author: string) => void;
  killGraph: () => void;
  emptySearch: boolean;
  loading: boolean;
  tooLarge: boolean;
  setTooLarge: React.Dispatch<React.SetStateAction<boolean>>;
}

const Graph: FunctionComponent<GraphProps> = ({
  dimensions,
  graphData,
  handleGraphExpand,
  authorDict,
  selectedAuthor,
  setSelectedAuthor,
  setSelectedArticle,
  removeSelectedAuthor,
  handleQuickSearch,
  killGraph,
  emptySearch,
  loading,
  tooLarge,
  setTooLarge,
}) => {
  const [emptyGraph, setEmptyGraph] = useState(true);

  const svgRef = useRef<SVGSVGElement>(null);

  useEffect(() => {
    if (!emptyGraph) {
      const svg = select<SVGSVGElement, GraphNode>(svgRef.current as any);
      drawGraph(svg, graphData as GraphData, dimensions, handleClick);
      graphData?.links && setEmptyGraph(false);
    }
  }, [graphData, dimensions]);

  const handleClick = (author: string) => {
    setSelectedAuthor(author);
  };

  const handleExpandClick = () => {
    handleGraphExpand(selectedAuthor);
  };

  const handleClean = () => {
    killGraph();
    setEmptyGraph(true);
    setTooLarge(false);
  };

  return (
    <div className="graph-container">
      <div className="data-container">
        <svg ref={svgRef} className="graph-svg"></svg>
      </div>
      <RightSidebar
        selectedAuthor={selectedAuthor}
        setSelectedAuthor={setSelectedAuthor}
        handleExpandClick={handleExpandClick}
        authorDict={authorDict}
        removeSelectedAuthor={removeSelectedAuthor}
        setSelectedArticle={setSelectedArticle}
      />
      <TinySearchBar handleQuickSearch={handleQuickSearch} />
      {graphData && Object.keys(graphData as GraphData).length > 0 && (
        <div className="graph-clean">
          <svg
            className="graph-icon-clean"
            viewBox="0 0 448 512"
            onClick={handleClean}
          >
            <defs>
              <linearGradient id="icon-red-gradient" x2="0.35" y2="1">
                <stop offset="0%" stopColor="var(--red-stop)" />
                <stop offset="30%" stopColor="var(--red-stop)" />
                <stop offset="100%" stopColor="var(--red-bot)" />
              </linearGradient>
            </defs>
            <path d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"></path>
          </svg>
          <p className="kill-text">Are you sure you want to clear graph?</p>
        </div>
      )}
      <GraphErrorHandler
        emptySearch={emptySearch}
        loading={loading}
        tooLarge={tooLarge}
        emptyGraph={emptyGraph}
      />
    </div>
  );
};

export default Graph;
