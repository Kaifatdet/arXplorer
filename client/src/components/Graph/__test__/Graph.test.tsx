import '@testing-library/jest-dom/extend-expect';
import * as React from 'react';
import {
  render,
  fireEvent,
  screen,
  cleanup,
  queryByText,
} from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Graph from '../Graph';
import { curveStepBefore } from 'd3';
import { mockData } from './mockData';

//getbytestid & provide svg testid: graph-svg to know if rendered
//need to create graph & expect nodes
//getbyquery & query circle
afterEach(cleanup);

const mockDimensions = { width: 1425, height: 714 };
describe('Graph', () => {
  beforeEach(() => {
    render(
      <Graph
        dimensions={mockDimensions}
        graphData={mockData}
        emptyGraph={false}
      />
    );
  });
  test('renders Graph component', () => {
    expect(screen.getByTestId('graph-svg')).toBeInTheDocument();
    screen.debug();
  });

  test('renders nodes', () => {
    const circle = document.querySelector('circle');
    fireEvent.click(circle); //vs user event
    //expect(screen.getByText).toBeTheDocument() after clicked on circle
    screen.debug();
  });
});

//unit tests for graph elements and drag helpers
//expect function to be output
//clickable or writable events
// const { getByTestId } = render(
//   <Graph dimensions={mockDimensions} graphData={mockData} />
//render Graph
//according to interface provide props
