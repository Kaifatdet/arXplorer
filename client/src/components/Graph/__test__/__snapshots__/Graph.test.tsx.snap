// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph renders nodes 1`] = `
<body>
  <div>
    <div
      class="graph-container"
    >
      <div
        class="data-container"
      >
        <svg
          class="graph-svg viewBox"
          data-testid="graph-svg"
          viewBox="0 0 1425 714"
        >
          <rect
            class="bg"
            fill="transparent"
            height="714"
            stroke="transparent"
            width="1425"
            x="0"
            y="0"
          />
          <g
            class="nodes"
          >
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
            <line
              stroke="#707070"
              stroke-opacity="0.6"
              stroke-width="1"
            />
          </g>
          <g
            class="nodes"
          >
            <circle
              fill="#a6cee3"
              r="8"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#1f78b4"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#1f78b4"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#a6cee3"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#a6cee3"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#a6cee3"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#a6cee3"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#b2df8a"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#b2df8a"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#33a02c"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#33a02c"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#fb9a99"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#fb9a99"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#e31a1c"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#a6cee3"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#a6cee3"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#a6cee3"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#a6cee3"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#a6cee3"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#a6cee3"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#a6cee3"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#a6cee3"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#a6cee3"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#a6cee3"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
            <circle
              fill="#a6cee3"
              r="5"
              stroke="black"
              stroke-width="0.3"
            />
          </g>
          <g
            class="nodes"
          >
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Darwin Chang
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Marc Schumann
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Laura Baudis
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              W. -Y. Keung
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Ernest Ma
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              A. Zee
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              R. N. Mohapatra
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Darwin Bautista
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Raimarc Dionido
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Oscar Darwin
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Stefan Kiefer
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Mladen Georgiev
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Jai Singh
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Dung-Hai Lee
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Wai-Yee Keung
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Ivan Phillips
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Jiang Liu
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Palash B. Pal
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Palash Pal
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Xiao-Gang He
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Sandip Pakvasa
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Shih-Chang Lee
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Paul Turcotte
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              Chuan-Hung Chen
            </text>
            <text
              class="label"
              font-size="1.8rem"
              style="fill: whitesmoke;"
              text-anchor="middle"
              visibility="hidden"
            >
              David Bowser-Chao
            </text>
          </g>
          <g
            class="legend"
          >
            <circle
              fill="#a6cee3"
              r="1rem"
              stroke="#000"
              stroke-width="0.3"
            />
            <circle
              fill="#1f78b4"
              r="1rem"
              stroke="#000"
              stroke-width="0.3"
            />
            <circle
              fill="#b2df8a"
              r="1rem"
              stroke="#000"
              stroke-width="0.3"
            />
            <circle
              fill="#33a02c"
              r="1rem"
              stroke="#000"
              stroke-width="0.3"
            />
            <circle
              fill="#fb9a99"
              r="1rem"
              stroke="#000"
              stroke-width="0.3"
            />
            <circle
              fill="#e31a1c"
              r="1rem"
              stroke="#000"
              stroke-width="0.3"
            />
          </g>
          <g
            class="legend"
          >
            <text
              class="legend-label"
              font-size="1.2rem"
              style="fill: whitesmoke;"
            >
              High Energy Physics - Phenomenology
            </text>
            <text
              class="legend-label"
              font-size="1.2rem"
              style="fill: whitesmoke;"
            >
              Instrumentation and Methods for Astrophysics
            </text>
            <text
              class="legend-label"
              font-size="1.2rem"
              style="fill: whitesmoke;"
            >
              Learning
            </text>
            <text
              class="legend-label"
              font-size="1.2rem"
              style="fill: whitesmoke;"
            >
              Logic in Computer Science
            </text>
            <text
              class="legend-label"
              font-size="1.2rem"
              style="fill: whitesmoke;"
            >
              Atmospheric and Oceanic Physics
            </text>
            <text
              class="legend-label"
              font-size="1.2rem"
              style="fill: whitesmoke;"
            >
              Strongly Correlated Electrons
            </text>
          </g>
          <g
            class="legend"
          >
            <circle
              fill="#a6cee3"
              r="5"
              stroke="#000"
              stroke-width="0.3"
            />
            <circle
              fill="#a6cee3"
              r="6.324555320336759"
              stroke="#000"
              stroke-width="0.3"
            />
            <circle
              fill="#a6cee3"
              r="10"
              stroke="#000"
              stroke-width="0.3"
            />
          </g>
          <g
            class="legend"
          >
            <text
              class="legend-label"
              font-size="1.2rem"
              style="fill: whitesmoke;"
            >
              1 collaborator
            </text>
            <text
              class="legend-label"
              font-size="1.2rem"
              style="fill: whitesmoke;"
            >
              10 collaborators
            </text>
            <text
              class="legend-label"
              font-size="1.2rem"
              style="fill: whitesmoke;"
            >
              25 collaborators
            </text>
          </g>
        </svg>
      </div>
      <div
        class="rsb-container hide-rsb"
      >
        <div
          class="rsb-icon"
        >
          <svg
            class="rsb-icon-close"
            viewBox="0 0 352 512"
          >
            <defs>
              <lineargradient
                id="icon-red-gradient"
                x2="0.35"
                y2="1"
              >
                <stop
                  offset="0%"
                  stop-color="var(--red-stop)"
                />
                <stop
                  offset="30%"
                  stop-color="var(--red-stop)"
                />
                <stop
                  offset="100%"
                  stop-color="var(--red-bot)"
                />
              </lineargradient>
            </defs>
            <path
              d="M242.72 256 l100.07-100.07 c12.28-12.28 12.28-32.19 0-44.48 l-22.24-22.24 c-12.28-12.28-32.19-12.28-44.48 0 L176 189.28 75.93 89.21 c-12.28-12.28-32.19-12.28-44.48 0 L9.21 111.45 c-12.28 12.28-12.28 32.19 0 44.48 L109.28 256 9.21 356.07 c-12.28 12.28-12.28 32.19 0 44.48 l22.24 22.24 c12.28 12.28 32.2 12.28 44.48 0 L176 322.72 l100.07 100.07 c12.28 12.28 32.2 12.28 44.48 0 l22.24-22.24 c12.28-12.28 12.28-32.19 0-44.48 L242.72 256 z"
            />
          </svg>
        </div>
        <div
          class="rsb-details"
        >
          <div
            class="rsb-author"
          />
          <div
            class="rsb-small-link"
          >
            show details
          </div>
        </div>
        <div
          class="rsb-list"
        >
          No articles to display
        </div>
        <div
          class="rsb-buttons-container"
        >
          <button
            class="rsb-expand-btn"
            type="submit"
          >
            Expand graph
          </button>
          <button
            class="rsb-delete-btn"
            type="submit"
          >
            Remove
          </button>
        </div>
      </div>
      <form
        class="tiny-container"
      >
        <label
          class="search-label"
          for="search-bar"
        >
          Quick Search
        </label>
        <input
          class="tinysearch-input"
          data-testid="tiny-search-bar"
          id="search-bar"
          name="tiny-search"
          placeholder="Search for author..."
          type="text"
          value=""
        />
        <button
          class="tinysearch-btn"
          data-testid="tiny-search-button"
          type="submit"
        >
          Quicksearch
        </button>
      </form>
      <div
        class="graph-clean"
      >
        <svg
          class="graph-icon-clean"
          viewBox="0 0 448 512"
        >
          <defs>
            <lineargradient
              id="icon-red-gradient"
              x2="0.35"
              y2="1"
            >
              <stop
                offset="0%"
                stop-color="var(--red-stop)"
              />
              <stop
                offset="30%"
                stop-color="var(--red-stop)"
              />
              <stop
                offset="100%"
                stop-color="var(--red-bot)"
              />
            </lineargradient>
          </defs>
          <path
            d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
          />
        </svg>
        <h1
          class="kill-text"
        >
          Are you sure you want to clear graph?
        </h1>
      </div>
    </div>
  </div>
</body>
`;
